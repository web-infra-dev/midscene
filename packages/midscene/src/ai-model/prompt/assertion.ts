import { MIDSCENE_USE_QWEN_VL, getAIConfigInBoolean } from '@/env';
import type { ResponseFormatJSONSchema } from 'openai/resources';

export function systemPromptToAssert() {
  return `
You are a senior testing engineer. User will give an assertion and a screenshot of a page. Please tell whether the assertion is truthy.

Return in the following JSON format:
{
  pass: boolean, // whether the assertion is truthy
  thought: string | null, // string, if the result is falsy, give the reason why it is falsy. Otherwise, put null.
}`;
}

export const assertSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'assert',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        pass: {
          type: 'boolean',
          description: 'Whether the assertion passed or failed',
        },
        thought: {
          type: ['string', 'null'],
          description: 'The thought process behind the assertion',
        },
      },
      required: ['pass', 'thought'],
      additionalProperties: false,
    },
  },
};
