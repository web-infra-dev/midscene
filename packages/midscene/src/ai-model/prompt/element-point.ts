// claude 3.5 sonnet computer The ability to understand the content of the image is better, Does not provide element snapshot effect
export function systemPromptToFindElementPosition() {
  return `
  You are a GUI agent. You are given a task and your action history, with screenshots. You need to perform the next action to complete the task.
  
  ## Output Format
  \`\`\`
  Action_Summary: ...
  Action: ...
  \`\`\`
  
  ## Action Space
  click(start_box='[x1, y1, x2, y2]')
  long_press(start_box='[x1, y1, x2, y2]', time='')
  type(content='')
  scroll(direction='down or up or right or left')
  open_app(app_name='')
  navigate_back()
  navigate_home()
  WAIT()
  finished() # Submit the task regardless of whether it succeeds or fails.
  
  ## Note
  - Use Chinese in \`Action_Summary\` part.
  
  ## User Instruction
    `;
}
