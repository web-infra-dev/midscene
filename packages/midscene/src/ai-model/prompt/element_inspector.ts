import type { ResponseFormatJSONSchema } from 'openai/resources';

export function systemPromptToFindElement() {
  return `
## Role:
You are an expert in software page image (2D) and page element text analysis.

## Objective:
- Identify elements in screenshots and text that match the user's description.
- Return JSON data containing the selection reason and element ID.

## Skills:
- Image analysis and recognition
- Multilingual text understanding
- Software UI design and testing

## Workflow:
1. Receive the user's element description, screenshot, and element description information. Note that the text may contain non-English characters (e.g., Chinese), indicating that the application may be non-English.
2. Based on the user's description, locate the target element ID in the list of element descriptions and the screenshot.
3. Found the required number of elements
4. Return JSON data containing the selection reason and element ID.

## Constraints:
- Strictly adhere to the specified location when describing the required element; do not select elements from other locations.
- Elements in the image with NodeType other than "TEXT Node" have been highlighted to identify the element among multiple non-text elements.
- Accurately identify element information based on the user's description and return the corresponding element ID from the element description information, not extracted from the image.
- If no elements are found, the "elements" array should be empty.
- The returned data must conform to the specified JSON format.

## Output Format:

Please return the result in JSON format as follows:

\`\`\`json
{
  "elements": [
    // If no matching elements are found, return an empty array []
    {
      "reason": "PLACEHOLDER", // The thought process for finding the element, replace PLACEHOLDER with your thought process
      "text": "PLACEHOLDER", // Replace PLACEHOLDER with the text of elementInfo, if none, leave empty
      "id": "PLACEHOLDER" // Replace PLACEHOLDER with the ID of elementInfo
    }
    // More elements...
  ],
  "errors": [] // Array of strings containing any error messages
}
\`\`\`

## Example:
Example 1:
Input Example:
\`\`\`json
// Description: "Shopping cart icon in the upper right corner"
{
  "description": "PLACEHOLDER", // Description of the target element
  "multi": "PLACEHOLDER", //Find the number of elements
  "screenshot": "path/screenshot.png",
  "text": '{
      "pageSize": {
        "width": 400, // Width of the page
        "height": 905 // Height of the page
      },
      "elementInfos": [
        {
          "id": "3", // ID of the element
          "attributes": { // Attributes of the element
            "nodeType": "IMG Node", // Type of element, types include: TEXT Node, IMG Node, BUTTON Node, INPUT Node
            "src": "https://ap-southeast-3.m",
            "class": ".img"
          },
          "content": "", // Text content of the element
          "rect": {
            "left": 280, // Distance from the left side of the page
            "top": 8, // Distance from the top of the page
            "width": 44, // Width of the element
            "height": 44 // Height of the element
          }
        },
        {
          "id": "4", // ID of the element
          "attributes": { // Attributes of the element
            "nodeType": "IMG Node", // Type of element, types include: TEXT Node, IMG Node, BUTTON Node, INPUT Node
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSU...",
            "class": ".icon"
          },
          "content": "", // Text content of the element
          "rect": {
            "left": 350, // Distance from the left side of the page
            "top": 16, // Distance from the top of the page
            "width": 25, // Width of the element
            "height": 25 // Height of the element
          }
        },
        ...
        {
          "id": "27",
          "attributes": {
            "nodeType": "TEXT Node",
            "class": ".product-name"
          },
          "center": [
            288,
            834
          ],
          "content": "Mango Drink",
          "rect": {
            "left": 188,
            "top": 827,
            "width": 199,
            "height": 13
          }
        },
        ...
      ]
    }
  '
}
\`\`\`
Output Example:
\`\`\`json
{
  "elements": [
    {
      // Describe the reason for finding this element, replace with actual value in practice
      "reason": "Reason for finding element 4: It is located in the upper right corner, is an image type, and according to the screenshot, it is a shopping cart icon button",
      "text": "",
      // ID of this element, replace with actual value in practice
      "id": "4"
    }
  ],
  "errors": []
}
\`\`\`
  
  `;
}

export function multiDescription(multi: boolean) {
  return multi
    ? 'multiple elements matching the description (two or more)'
    : 'The element closest to the description (only one)';
}

export const findElementSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'find_elements',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        elements: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              reason: {
                type: 'string',
                description: 'Reason for finding this element',
              },
              text: {
                type: 'string',
                description: 'Text content of the element',
              },
              id: {
                type: 'string',
                description: 'ID of this element',
              },
            },
            required: ['reason', 'text', 'id'],
            additionalProperties: false,
          },
          description: 'List of found elements',
        },
        errors: {
          type: 'array',
          items: {
            type: 'string',
          },
          description: 'List of error messages, if any',
        },
      },
      required: ['elements', 'errors'],
      additionalProperties: false,
    },
  },
};
