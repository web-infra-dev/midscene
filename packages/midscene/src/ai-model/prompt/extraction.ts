import { PromptTemplate } from '@langchain/core/prompts';
import type { ResponseFormatJSONSchema } from 'openai/resources';

export function systemPromptToExtract() {
  return `
You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.

The user will give you a screenshot, the contents of it (optional), and some data requirements in DATA_DEMAND. You need to extract the data according to the DATA_DEMAND.

Return in the following JSON format:
{
  data: any, // the extracted data from extract_data_from_UI skill. Make sure both the value and scheme meet the DATA_DEMAND. If you want to write some description in this field, use the same language as the DATA_DEMAND.
  errors: [], // string[], error message if any
}
`;
}

export const extractDataPrompt = new PromptTemplate({
  template: `
pageDescription: {pageDescription}

Use your extract_data_from_UI skill to find the following data, placing it in the \`data\` field
DATA_DEMAND start:
=====================================
{dataKeys}

{dataQuery}
=====================================
DATA_DEMAND ends.
  `,
  inputVariables: ['pageDescription', 'dataKeys', 'dataQuery'],
});

export const extractDataSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'extract_data',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        data: {
          type: 'object',
          description: 'The extracted data from extract_data_from_UI skill',
        },
        errors: {
          type: 'array',
          items: {
            type: 'string',
          },
          description: 'Error messages, if any',
        },
      },
      required: ['data', 'errors'],
      additionalProperties: false,
    },
  },
};
