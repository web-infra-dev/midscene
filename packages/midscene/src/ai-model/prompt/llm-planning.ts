import { MATCH_BY_POSITION, getAIConfig } from '@/env';
import { PromptTemplate } from '@langchain/core/prompts';
import type { ResponseFormatJSONSchema } from 'openai/resources';
import { samplePageDescription } from './util';

const quickAnswerFormat = () => {
  const matchByPosition = getAIConfig(MATCH_BY_POSITION);

  const locationFormat = {
    position: {
      description:
        '"position": { x: number; y: number } // Represents the position of the element; replace with actual values in practice (ensure it reflects the element\'s position)',
      format: '"position": { x: number; y: number }',
      sample:
        '{"prompt": "the search bar" // Use language consistent with the information on the page}',
      locateParam: `{
        "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
      } | null // If it's not on the page, the LocateParam should be null`,
    },
    id: {
      description:
        '"id": string // Represents the ID of the element; replace with actual values in practice',
      format: '"id": string',
      sample: `{"id": "c81c4e9a33", "prompt": "the search bar"}`,
      locateParam: `{
        "id": string, // the id of the element found. It should either be the id marked with a rectangle in the screenshot or the id described in the description.
        "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
      } | null // If it's not on the page, the LocateParam should be null`,
    },
  };

  const type = matchByPosition ? 'position' : 'id';
  const format = locationFormat[type];

  return {
    description: format.description,
    format: format.format,
    sample: format.sample,
    locateParam: format.locateParam,
  };
};

const systemTemplate = `
## Role

You are a versatile professional in software UI automation. Your outstanding contributions will impact the user experience of billions of users.

## Objective

- Decompose the instruction user asked into a series of actions
- Locate the target element if possible
- If the instruction cannot be accomplished, give a further plan.

## Workflow

1. Receive the user's element description, screenshot, and instruction.
2. Decompose the user's task into a sequence of actions, and place it in the \`actions\` field. There are different types of actions (Tap / Hover / Input / KeyboardPress / Scroll / FalsyConditionStatement / Sleep). The "About the action" section below will give you more details.
3. Precisely locate the target element if it's already shown in the screenshot, put the location info in the \`locate\` field of the action.
4. If some target elements is not shown in the screenshot, consider the user's instruction is not feasible on this page. Follow the next steps.
5. Consider whether the user's instruction will be accomplished after all the actions
 - If yes, set \`taskWillBeAccomplished\` to true
 - If no, don't plan more actions by closing the array. Get ready to reevaluate the task. Some talent people like you will handle this. Give him a clear description of what have been done and what to do next. Put your new plan in the \`furtherPlan\` field. The "How to compose the \`taskWillBeAccomplished\` and \`furtherPlan\` fields" section will give you more details.

## Constraints

- All the actions you composed MUST be based on the page context information you get.
- Trust the "What have been done" field about the task (if any), don't repeat actions in it.
- Respond only with valid JSON. Do not write an introduction or summary or markdown prefix like \`\`\`json\`.
- If you cannot plan any action at all (i.e. empty actions array), set reason in the \`error\` field.

## About the \`actions\` field

### The common \`locate\` param

The \`locate\` param is commonly used in the \`param\` field of the action, means to locate the target element to perform the action, it follows the following scheme:

type LocateParam = {locateParam}

### Supported actions

Each action has a \`type\` and corresponding \`param\`. To be detailed:
- type: 'Tap', tap the located element
  * {{ locate: {sample}, param: null }}
- type: 'Hover', move mouse over to the located element
  * {{ locate: LocateParam, param: null }}
- type: 'Input', replace the value in the input field
  * {{ locate: LocateParam, param: {{ value: string }} }}
  * \`value\` is the final required input value based on the existing input. No matter what modifications are required, just provide the final value to replace the existing input value. 
- type: 'KeyboardPress', press a key
  * {{ param: {{ value: string }} }}
- type: 'Scroll', scroll up or down.
  * {{ 
      locate: LocateParam | null, 
      param: {{ 
        direction: 'down'(default) | 'up' | 'right' | 'left', 
        scrollType: 'once' (default) | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft', 
        distance: null | number 
      }} 
    }}
    * To scroll some specific element, put the element at the center of the region in the \`locate\` field. If it's a page scroll, put \`null\` in the \`locate\` field. 
    * \`param\` is required in this action. If some fields are not specified, use direction \`down\`, \`once\` scroll type, and \`null\` distance.
- type: 'FalsyConditionStatement'
  * {{ param: null }}
  * use this action when the instruction is an "if" statement and the condition is falsy.
- type: 'Sleep'
  * {{ param: {{ timeMs: number }} }}

## How to compose the \`taskWillBeAccomplished\` and \`furtherPlan\` fields ?

\`taskWillBeAccomplished\` is a boolean field, means whether the task will be accomplished after all the actions.

\`furtherPlan\` is used when the task cannot be accomplished. It follows the scheme {{ whatHaveDone: string, whatToDoNext: string }}:
- \`whatHaveDone\`: a string, describe what have been done after the previous actions.
- \`whatToDoNext\`: a string, describe what should be done next after the previous actions has finished. It should be a concise and clear description of the actions to be performed. Make sure you don't lose any necessary steps user asked.
`;

const outputTemplate = `
## Output JSON Format:

The JSON format is as follows:

{{
  "actions": [
    {{
      "thought": "Reasons for generating this task, and why this task is feasible on this page.", // Use the same language as the user's instruction.
      "type": "Tap",
      "param": null,
      "locate": {sample} | null,
    }},
    // ... more actions
  ],
  "taskWillBeAccomplished": boolean,
  "furtherPlan": {{ "whatHaveDone": string, "whatToDoNext": string }} | null, // Use the same language as the user's instruction.
  "error"?: string // Use the same language as the user's instruction.
}}
Here is an example of how to decompose a task:

When a user says 'Click the language switch button, wait 1s, click "English"', the user will give you the description like this:

====================
{pageDescription}
====================

By viewing the page screenshot and description, you should consider this and output the JSON:

* The main steps should be: tap the switch button, sleep, and tap the 'English' option 
* The language switch button is shown in the screenshot, but it's not marked with a rectangle. So we have to use the page description to find the element. By carefully checking the context information (coordinates, attributes, content, etc.), you can find the element.
* The "English" option button is not shown in the screenshot now, it means it may only show after the previous actions are finished. So the last action will have a \`null\` value in the \`locate\` field. 
* The task cannot be accomplished (because we cannot see the "English" option now), so a \`furtherPlan\` field is needed.

{{
  "actions":[
    {{
      "type": "Tap", 
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {sample},
    }},
    {{
      "type": "Sleep",
      "thought": "Wait for 1 second to ensure the language options are displayed.",
      "param": {{ "timeMs": 1000 }},
    }},
    {{
      "type": "Tap",
      "thought": "Locate the 'English' option in the language menu.",
      "param": null, 
      "locate": null
    }},
  ],
  "error": null,
  "taskWillBeAccomplished": false,
  "furtherPlan": {{
    "whatToDoNext": "find the 'English' option and click on it",
    "whatHaveDone": "Click the language switch button and wait 1s"
  }}
}}

Here is another example of how to tolerate error situations only when the instruction is an "if" statement:

If the user says "If there is a popup, close it", you should consider this and output the JSON:

* By viewing the page screenshot and description, you cannot find the popup, so the condition is falsy.
* The instruction itself is an "if" statement, it means the user can tolerate this situation, so you should leave a \`FalsyConditionStatement\` action.

{{
  "actions": [{{
      "type": "FalsyConditionStatement",
      "thought": "There is no popup on the page",
      "param": null
    }}
  ],
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}}

For contrast, if the user says "Close the popup" in this situation, you should consider this and output the JSON:

{{
  "actions": [],
  "error": "The instruction and page context are irrelevant, there is no popup on the page",
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}}

Here is an example of when task is accomplished, don't plan more actions:

When the user ask to "Wait 4s", you should consider this:

{{
  "actions": [
    {{
      "type": "Sleep",
      "thought": "Wait for 4 seconds",
      "param": {{ "timeMs": 4000 }},
    }},
  ],
  "taskWillBeAccomplished": true,
  "furtherPlan": null // All steps have been included in the actions, so no further plan is needed
}}

Here is an example of what NOT to do:

Wrong output:

{{
  "actions":[
    {{
      "type": "Tap",
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {{
        {sample}, // WRONG:prompt is missing
      }}
    }},
    {{
      "type": "Tap", 
      "thought": "Click the English option",
      "param": null,
      "locate": null, // This means the 'English' option is not shown in the screenshot, the task cannot be accomplished
    }}
  ],
  "taskWillBeAccomplished": false,
  // WRONG: should not be null
  "furtherPlan": null,
}}

Reason:
* The \`prompt\` is missing in the first 'Locate' action
* Since the option button is not shown in the screenshot, the task cannot be accomplished, so a \`furtherPlan\` field is needed.
`;

export async function systemPromptToTaskPlanning() {
  const promptTemplate = new PromptTemplate({
    template: `${systemTemplate}\n\n${outputTemplate}`,
    inputVariables: ['pageDescription', 'sample', 'locateParam'],
  });

  return await promptTemplate.format({
    pageDescription: samplePageDescription,
    sample: quickAnswerFormat().sample,
    locateParam: quickAnswerFormat().locateParam,
  });
}

export const planSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'action_items',
    strict: true,
    schema: {
      type: 'object',
      strict: true,
      properties: {
        actions: {
          type: 'array',
          items: {
            type: 'object',
            strict: true,
            properties: {
              thought: {
                type: 'string',
                description:
                  'Reasons for generating this task, and why this task is feasible on this page',
              },
              type: {
                type: 'string',
                description: 'Type of action, like "Tap", "Hover", etc.',
              },
              param: {
                type: ['object', 'null'],
                description:
                  'Parameter of the action, can be null ONLY when the type field is Tap or Hover',
                additionalProperties: false,
              },
              locate: {
                type: ['object', 'null'],
                properties: {
                  ...(getAIConfig(MATCH_BY_POSITION)
                    ? {
                        position: {
                          type: 'object',
                          properties: {
                            x: { type: 'number' },
                            y: { type: 'number' },
                          },
                          required: ['x', 'y'],
                          additionalProperties: false,
                        },
                      }
                    : {
                        id: { type: 'string' },
                      }),
                  prompt: { type: 'string' },
                },
                required: [
                  getAIConfig(MATCH_BY_POSITION) ? 'position' : 'id',
                  'prompt',
                ],
                additionalProperties: false,
                description: 'Location information for the target element',
              },
            },
            required: ['thought', 'type', 'param', 'locate'],
            additionalProperties: false,
          },
          description: 'List of actions to be performed',
        },
        taskWillBeAccomplished: {
          type: 'boolean',
          description:
            'Whether the task will be accomplished after the actions',
        },
        furtherPlan: {
          type: ['object', 'null'],
          properties: {
            whatHaveDone: { type: 'string' },
            whatToDoNext: { type: 'string' },
          },
          required: ['whatHaveDone', 'whatToDoNext'],
          additionalProperties: false,
          description: 'Plan the task when the task cannot be accomplished',
        },
        error: {
          type: ['string', 'null'],
          description: 'Overall error messages',
        },
      },
      required: ['actions', 'taskWillBeAccomplished', 'furtherPlan', 'error'],
      additionalProperties: false,
    },
  },
};

export const generateTaskBackgroundContext = (
  userPrompt: string,
  originalPrompt?: string,
  whatHaveDone?: string,
) => {
  if (originalPrompt && whatHaveDone) {
    return `
    Here is the instruction:
    =====================================
    ${userPrompt}
    =====================================
    
    For your information, this is a task that some important person handed to you. Here is the original task description and what have been done after the previous actions:
    =====================================
    Original task description: ${originalPrompt}
    =====================================
    What have been done: ${whatHaveDone}
    =====================================
    `;
  }

  return `
  Here is the instruction:
  =====================================
  ${userPrompt}
  =====================================
  `;
};

export const automationUserPrompt = new PromptTemplate({
  template: `
    pageDescription:
    =====================================
    {pageDescription}
    =====================================

    {taskBackgroundContext}
  `,
  inputVariables: ['pageDescription', 'taskBackgroundContext'],
});
