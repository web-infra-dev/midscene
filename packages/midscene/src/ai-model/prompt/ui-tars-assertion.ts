import { getTimeZoneInfo } from './ui-tars-planning';

export const language = getTimeZoneInfo().isChina ? 'Chinese' : 'English';

/**
 * UI Tars model doesn't support response_format, this is a stable workaround to forcefully ask UI-Tars generate Json Markdown as reponse
 *
 * Tip: It requires to set Temporature to 0
 */
export function systemPromptToAssertForUITars() {
  return `
You are a senior testing engineer. User will give an assertion and a screenshot of a page. Please tell whether the assertion is truthy.

## Output Json String Format
\`\`\`
"{
  "pass": <<is a boolean value from the enum [true, false], true means the assertion is truthy>>, 
  "thought": "<<is a string, give the reason why the assertion is falsy or truthy. Otherwise.>>"
}"
\`\`\`

## Rules **MUST** follow
- Make sure to return **only** the JSON, with **no additional** text or explanations.
- Use ${language} in \`thought\` part.
- You **MUST** strictly follow up the **Output Json String Format**.
`;
}
