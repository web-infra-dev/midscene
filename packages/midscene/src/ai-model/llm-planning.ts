import assert from 'node:assert';
import { MIDSCENE_USE_QWEN_VL, getAIConfigInBoolean } from '@/env';
import type {
  PlanningAIResponse,
  PlanningLocateParam,
  UIContext,
} from '@/types';
import { AIActionType, type AIArgs, callAiFn } from './common';
import {
  automationUserPrompt,
  generateTaskBackgroundContext,
  systemPromptToTaskPlanning,
} from './prompt/llm-planning';
import { describeUserPage } from './prompt/util';

// transform the param of locate from qwen mode
export function fillLocateParam(locate: PlanningLocateParam) {
  if (locate?.bbox_2d && !locate?.bbox) {
    locate.bbox = locate.bbox_2d;
    // biome-ignore lint/performance/noDelete: <explanation>
    delete locate.bbox_2d;
  }

  const defaultBboxSize = 10;
  if (locate?.bbox) {
    locate.bbox[0] = Math.round(locate.bbox[0]);
    locate.bbox[1] = Math.round(locate.bbox[1]);
    locate.bbox[2] =
      typeof locate.bbox[2] === 'number'
        ? Math.round(locate.bbox[2])
        : Math.round(locate.bbox[0] + defaultBboxSize);
    locate.bbox[3] =
      typeof locate.bbox[3] === 'number'
        ? Math.round(locate.bbox[3])
        : Math.round(locate.bbox[1] + defaultBboxSize);
  }

  return locate;
}

export async function plan(
  userInstruction: string,
  opts: {
    log?: string;
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
): Promise<PlanningAIResponse> {
  const { callAI, context } = opts || {};
  const { screenshotBase64, screenshotBase64WithElementMarker, size } = context;
  const { description: pageDescription } = await describeUserPage(context);

  const systemPrompt = await systemPromptToTaskPlanning();
  const taskBackgroundContextText = generateTaskBackgroundContext(
    userInstruction,
    opts.log,
  );
  const userInstructionPrompt = await automationUserPrompt().format({
    pageDescription,
    taskBackgroundContext: taskBackgroundContextText,
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64WithElementMarker || screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const call = callAI || callAiFn;
  const { content, usage } = await call(msgs, AIActionType.PLAN);
  const rawResponse = JSON.stringify(content, undefined, 2);
  const planFromAI = content;
  const actions =
    (planFromAI.action ? [planFromAI.action] : planFromAI.actions) || [];
  const returnValue: PlanningAIResponse = {
    ...planFromAI,
    actions,
    rawResponse,
    usage,
  };

  if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    actions.forEach((action) => {
      if (action.locate) {
        action.locate = fillLocateParam(action.locate);
      }
    });
  }

  assert(planFromAI, "can't get plans from AI");
  assert(
    actions.length > 0 || returnValue.finish,
    `Failed to plan actions: ${planFromAI.error || '(no error details)'}`,
  );

  return returnValue;
}
