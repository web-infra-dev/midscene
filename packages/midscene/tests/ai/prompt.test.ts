import { systemPromptToTaskPlanning } from '@/ai-model/prompt/llm-planning';
import { describe, expect, it, test } from 'vitest';

describe('automation - computer', () => {
  it('should be able to generate prompt', async () => {
    const prompt = await systemPromptToTaskPlanning();
    expect(prompt).toBeDefined();
    expect(prompt).toMatchSnapshot();
  });
});
