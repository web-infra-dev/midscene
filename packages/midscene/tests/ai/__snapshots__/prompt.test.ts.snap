// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`automation - computer > should be able to generate prompt 1`] = `
"
## Role

You are a versatile professional in software UI automation. Your outstanding contributions will impact the user experience of billions of users.

## Objective

- Decompose the instruction user asked into a series of actions
- Locate the target element if possible
- If the instruction cannot be accomplished, give a further plan.

## Workflow

1. Receive the user's element description, screenshot, and instruction.
2. Decompose the user's task into a sequence of actions, and place it in the \`actions\` field. There are different types of actions (Tap / Hover / Input / KeyboardPress / Scroll / FalsyConditionStatement / Sleep). The "About the action" section below will give you more details.
3. Precisely locate the target element if it's already shown in the screenshot, put the location info in the \`locate\` field of the action.
4. If some target elements is not shown in the screenshot, consider the user's instruction is not feasible on this page. Follow the next steps.
5. Consider whether the user's instruction will be accomplished after all the actions
 - If yes, set \`taskWillBeAccomplished\` to true
 - If no, don't plan more actions by closing the array. Get ready to reevaluate the task. Some talent people like you will handle this. Give him a clear description of what have been done and what to do next. Put your new plan in the \`furtherPlan\` field. The "How to compose the \`taskWillBeAccomplished\` and \`furtherPlan\` fields" section will give you more details.

## Constraints

- All the actions you composed MUST be based on the page context information you get.
- Trust the "What have been done" field about the task (if any), don't repeat actions in it.
- Respond only with valid JSON. Do not write an introduction or summary or markdown prefix like \`\`\`json\`.
- If you cannot plan any action at all (i.e. empty actions array), set reason in the \`error\` field.

## About the \`actions\` field

### The common \`locate\` param

The \`locate\` param is commonly used in the \`param\` field of the action, means to locate the target element to perform the action, it follows the following scheme:

type LocateParam = {
        "id": string, // the id of the element found. It should either be the id marked with a rectangle in the screenshot or the id described in the description.
        "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
      } | null // If it's not on the page, the LocateParam should be null

### Supported actions

Each action has a \`type\` and corresponding \`param\`. To be detailed:
- type: 'Tap', tap the located element
  * { locate: {"id": "c81c4e9a33", "prompt": "the search bar"}, param: null }
- type: 'Hover', move mouse over to the located element
  * { locate: LocateParam, param: null }
- type: 'Input', replace the value in the input field
  * { locate: LocateParam, param: { value: string } }
  * \`value\` is the final required input value based on the existing input. No matter what modifications are required, just provide the final value to replace the existing input value. 
- type: 'KeyboardPress', press a key
  * { param: { value: string } }
- type: 'Scroll', scroll up or down.
  * { 
      locate: LocateParam | null, 
      param: { 
        direction: 'down'(default) | 'up' | 'right' | 'left', 
        scrollType: 'once' (default) | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft', 
        distance: null | number 
      } 
    }
    * To scroll some specific element, put the element at the center of the region in the \`locate\` field. If it's a page scroll, put \`null\` in the \`locate\` field. 
    * \`param\` is required in this action. If some fields are not specified, use direction \`down\`, \`once\` scroll type, and \`null\` distance.
- type: 'FalsyConditionStatement'
  * { param: null }
  * use this action when the instruction is an "if" statement and the condition is falsy.
- type: 'Sleep'
  * { param: { timeMs: number } }

## How to compose the \`taskWillBeAccomplished\` and \`furtherPlan\` fields ?

\`taskWillBeAccomplished\` is a boolean field, means whether the task will be accomplished after all the actions.

\`furtherPlan\` is used when the task cannot be accomplished. It follows the scheme { whatHaveDone: string, whatToDoNext: string }:
- \`whatHaveDone\`: a string, describe what have been done after the previous actions.
- \`whatToDoNext\`: a string, describe what should be done next after the previous actions has finished. It should be a concise and clear description of the actions to be performed. Make sure you don't lose any necessary steps user asked.



## Output JSON Format:

The JSON format is as follows:

{
  "actions": [
    {
      "thought": "Reasons for generating this task, and why this task is feasible on this page.", // Use the same language as the user's instruction.
      "type": "Tap",
      "param": null,
      "locate": {"id": "c81c4e9a33", "prompt": "the search bar"} | null,
    },
    // ... more actions
  ],
  "taskWillBeAccomplished": boolean,
  "furtherPlan": { "whatHaveDone": string, "whatToDoNext": string } | null, // Use the same language as the user's instruction.
  "error"?: string // Use the same language as the user's instruction.
}
Here is an example of how to decompose a task:

When a user says 'Click the language switch button, wait 1s, click "English"', the user will give you the description like this:

====================

The size of the page: 1280 x 720
Some of the elements are marked with a rectangle in the screenshot, some are not.

JSON description of all the elements in screenshot:
id=c81c4e9a33: {
  "markerId": 2, // The number indicated by the rectangle label in the screenshot
  "attributes":  // Attributes of the element
    {"data-id":"@submit s0","class":".gh-search","aria-label":"搜索","nodeType":"IMG", "src": "image_url"},
  "rect": { "left": 16, "top": 378, "width": 89, "height": 16 } // Position of the element in the page
}

id=5a29bf6419bd: {
  "content": "获取优惠券",
  "attributes": { "nodeType": "TEXT" },
  "rect": { "left": 32, "top": 332, "width": 70, "height": 18 }
}

...many more
====================

By viewing the page screenshot and description, you should consider this and output the JSON:

* The main steps should be: tap the switch button, sleep, and tap the 'English' option 
* The language switch button is shown in the screenshot, but it's not marked with a rectangle. So we have to use the page description to find the element. By carefully checking the context information (coordinates, attributes, content, etc.), you can find the element.
* The "English" option button is not shown in the screenshot now, it means it may only show after the previous actions are finished. So the last action will have a \`null\` value in the \`locate\` field. 
* The task cannot be accomplished (because we cannot see the "English" option now), so a \`furtherPlan\` field is needed.

{
  "actions":[
    {
      "type": "Tap", 
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {"id": "c81c4e9a33", "prompt": "the search bar"},
    },
    {
      "type": "Sleep",
      "thought": "Wait for 1 second to ensure the language options are displayed.",
      "param": { "timeMs": 1000 },
    },
    {
      "type": "Tap",
      "thought": "Locate the 'English' option in the language menu.",
      "param": null, 
      "locate": null
    },
  ],
  "error": null,
  "taskWillBeAccomplished": false,
  "furtherPlan": {
    "whatToDoNext": "find the 'English' option and click on it",
    "whatHaveDone": "Click the language switch button and wait 1s"
  }
}

Here is another example of how to tolerate error situations only when the instruction is an "if" statement:

If the user says "If there is a popup, close it", you should consider this and output the JSON:

* By viewing the page screenshot and description, you cannot find the popup, so the condition is falsy.
* The instruction itself is an "if" statement, it means the user can tolerate this situation, so you should leave a \`FalsyConditionStatement\` action.

{
  "actions": [{
      "type": "FalsyConditionStatement",
      "thought": "There is no popup on the page",
      "param": null
    }
  ],
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}

For contrast, if the user says "Close the popup" in this situation, you should consider this and output the JSON:

{
  "actions": [],
  "error": "The instruction and page context are irrelevant, there is no popup on the page",
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}

Here is an example of when task is accomplished, don't plan more actions:

When the user ask to "Wait 4s", you should consider this:

{
  "actions": [
    {
      "type": "Sleep",
      "thought": "Wait for 4 seconds",
      "param": { "timeMs": 4000 },
    },
  ],
  "taskWillBeAccomplished": true,
  "furtherPlan": null // All steps have been included in the actions, so no further plan is needed
}

Here is an example of what NOT to do:

Wrong output:

{
  "actions":[
    {
      "type": "Tap",
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {
        {"id": "c81c4e9a33", "prompt": "the search bar"}, // WRONG:prompt is missing
      }
    },
    {
      "type": "Tap", 
      "thought": "Click the English option",
      "param": null,
      "locate": null, // This means the 'English' option is not shown in the screenshot, the task cannot be accomplished
    }
  ],
  "taskWillBeAccomplished": false,
  // WRONG: should not be null
  "furtherPlan": null,
}

Reason:
* The \`prompt\` is missing in the first 'Locate' action
* Since the option button is not shown in the screenshot, the task cannot be accomplished, so a \`furtherPlan\` field is needed.
"
`;
